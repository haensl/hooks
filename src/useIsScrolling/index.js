import { useState, useCallback } from 'react';
import { useDebounce, useIsMounted, useOnScroll } from '../';

const useIsScrolling = (
  el = window,
  scrollEndMs = 100
) => {
  const [isScrolling, setIsScrolling] = useState(false);
  const isMounted = useIsMounted();
  const onScroll = useCallback(() => {
    setIsScrolling(true);
  }, []);
  const onScrollEnd = useCallback(() => {
    if (isMounted()) {
      setIsScrolling(false);
    }
  }, [isMounted]);

  useOnScroll(onScroll, el);
  useOnScroll(useDebounce(onScrollEnd, scrollEndMs), el);

  return isScrolling;
};

export default useIsScrolling;
