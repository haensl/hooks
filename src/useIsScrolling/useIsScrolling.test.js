import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import useIsScrolling from './';

describe('useIsScrolling', () => {
  let TestComponent;
  let testComponent;
  let isScrolling;

  beforeAll(() => {
    jest.useFakeTimers();
    TestComponent = () => {
      isScrolling = useIsScrolling();

      if (typeof isScrolling === 'boolean') {
        return (<span>isScrolling: {
          (isScrolling && 'true') || 'false'
        }</span>);
      }

      return (<span>no is scrolling</span>);
    };
  });

  it('renders without crashing', () => {
    expect(shallow.bind(shallow, <TestComponent />))
      .not.toThrow();
  });

  it('renders as expected', () => {
    expect(shallow(<TestComponent />))
      .toMatchSnapshot();
  });

  describe('when the window is scrolled', () => {
    beforeAll(() => {
      testComponent = mount(<TestComponent />);
    });

    beforeEach(() => {
      act(() => {
        window.scrollY = 50;
        window.dispatchEvent(new Event('scroll'));
        testComponent.update();
      });
    });

    it('indicates that the window is being scrolled', () => {
      expect(isScrolling).toBe(true);
    });

    it('renders as expected', () => {
      expect(testComponent)
        .toMatchSnapshot();
    });

    describe('and scrolling has finished', () => {
      beforeEach(() => {
        act(() => {
          jest.advanceTimersByTime(100);
          testComponent.update();
        });
      });

      it('indicates that the window has stopped scrolling', () => {
        expect(isScrolling).toBe(false);
      });

      it('renders as expected', () => {
        expect(testComponent)
          .toMatchSnapshot();
      });
    });
  });
});
