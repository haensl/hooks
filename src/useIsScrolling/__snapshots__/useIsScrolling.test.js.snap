// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useIsScrolling renders as expected 1`] = `
<span>
  isScrolling: 
  false
</span>
`;

exports[`useIsScrolling when the window is scrolled and scrolling has finished renders as expected 1`] = `
<TestComponent>
  <span>
    isScrolling: 
    true
  </span>
</TestComponent>
`;

exports[`useIsScrolling when the window is scrolled renders as expected 1`] = `
<TestComponent>
  <span>
    isScrolling: 
    true
  </span>
</TestComponent>
`;
